Add a PUBLISH message for publisher initiated subscriptions (v2) (#6)
afrind/moq-transport: draft-ietf-moq-transport.md

the tracks does not depend on those parameters; this is in contrast to
protocols like HTTP, where request headers can alter the server response.

> ### Subscriptions

Reviewer: I thought it'd make sense to put this later in the SUBSCRIBE/FETCH section unless there's a reason it's under "Track"?

A subscription can be initiated by either a publisher or a subscriber.  A
publisher initiates a subscription to a track by sending the PUBLISH message.
The subscriber either accepts or rejects the subscription using PUBLISH_OK or
PUBLISH_ERROR.  A subscriber initiates a subscription to a track by sending the
SUBSCRIBE message.  The publisher either accepts or rejects the subscription
using SUBSCRIBE_OK or SUBSCRIBE_ERROR.  Once either of these sequences is

> complete, the subscription can be updated by the subscriber using

Reviewer 2: s/Once either of these sequences is complete/Once the subscription is successful,/...

> complete, the subscription can be updated by the subscriber using

Author: ```suggestion
successful, the subscription can be updated by the subscriber using
```

SUBSCRIBE_UPDATE, terminated by the subscriber using UNSUBSCRIBE, or terminated
by the publisher using SUBSCRIBE_DONE.

> All subscriptions have a Forward State which is either 0 or 1.  If the Forward

Reviewer 3: nit ... rather see this as true or false then described in message encoding how that gets mapped to bits

State is 0, the publisher does not send objects for the subscription.  If the
Forward State is 1, the publisher sends objects.  The initiator of the
subscription sets the initial Forward State in either PUBLISH or SUBSCRIBE.  The
sender of PUBLISH_OK can update the Forward State based on its preference.  Once
established, the subscriber can change the Forward State by sending
SUBSCRIBE_UPDATE.

> Either endpoint can initiate a subscription to a track without exhanging any

Reviewer: ```suggestion
Either endpoint can initiate a subscription to a track without exchanging any
```

> prior messages other than SETUP.  Relays however do not send any PUBLISH

Reviewer: SHOULD NOT?  Clearly they can.

Author: It's funny because we had the same exchange on #3 and I forgot to move it to this PR:

https://github.com/afrind/moq-transport/pull/3/files#r2061015708

> prior messages other than SETUP.  Relays however do not send any PUBLISH
> messages without first receiving a SUBSCRIBE_ANNOUNCES to a relevant track
> namespace or namespace prefix.

Author: ```suggestion
prior messages other than SETUP.  Relays MUST NOT send any PUBLISH messages
without knowing the client is interested in and authorized to receive the
content. The communication of intent and authorization can be accomplished by
the client sending SUBSCRIBE_ANNOUNCES, or conveyed in other mechanisms out of
band.
```

Discovery of MoQT servers is always done out-of-band. Namespace discovery can be
done in the context of an established MoQT session.

Given sufficient out of band information, it is valid for a subscriber to send a
SUBSCRIBE or FETCH message to a publisher (including a relay) without any
previous MoQT messages besides SETUP. However, SUBSCRIBE_ANNOUNCES, PUBLISH and
ANNOUNCE messages provide an in-band means of discovery of publishers for a
namespace.

The syntax of these messages is described in {{message}}.

## Subscribing to Namespaces

If the subscriber is aware of a namespace of interest, it can send

> SUBSCRIBE_ANNOUNCES to publishers/relays it has established a session with. The

Reviewer 2: Should we rename to Subscribe_Namespaces ?

Author: Ian asked me not to because it bloats the diff and makes it hard to read.  We will follow up, either by renaming, or moving to SUBSCRIBE.

forward the result to the application, so the application can decide which other
publishers to contact, if any.

An UNSUBSCRIBE_ANNOUNCES withdraws a previous SUBSCRIBE_ANNOUNCES. It does not

> prohibit the publishers from sending further ANNOUNCE or PUBLISH messages, but

Reviewer 2: ```suggestion
prohibit the original publishers from sending further ANNOUNCE or PUBLISH messages, but
```

> it does prevent relays from sending them.

Author: ```suggestion
prohibit original publishers from sending further ANNOUNCE or PUBLISH messages,
but it does prevent relays from sending them.
```

> but it does prevent relays from sending them.

Reviewer: Howso?

> but it does prevent relays from sending them.

Author: ```suggestion
but relays MUST not send a PUBLISH message to a client without knowing the
client is interested in and authorized to receive the content.
```

SHOULD close the session with a protocol error if it receives more than one.

An UNANNOUNCE message withdraws a previous ANNOUNCE, although it is not a
protocol error for the subscriber to send a SUBSCRIBE or FETCH message for a

> track in a namespace after receiving an UNANNOUNCE.

Reviewer 3: I'm not this is correct. (or it depends on other details )

 Imagine a client that is doing handover from cell to wifi. It was connected on cell, announced, there are subscribers and it is sending stuff.  Now it does an announce on wifi connection and unannounce of same track on cell connection. At all points it time, here it would be fine to send a relay the fetch. Perhaps this depends on how the announce and unannouce get aggregated by relay. 

Ignoring the above issue ... we certainly can have glare condition where the fetch was send before the unnanouce was received but the unnanouce was sent before the fetch was received. That' can't cause a protocol error.  So I think we need an error code for this.

Author: I didn't change this text (much) in this PR, so maybe you can file a separate issue -- it's not directly related to PUBLISH.  I think the reason the text is like this is: we say that it's OK to send a SUBSCRIBE or FETCH if you have never seen an ANNOUNCE, so it seems ANNOUNCE then UNANNOUNCE should put you in the same state.  General purpose relays won't do this, but maybe an application specific relay with an out of band configuration could?

In your example, the FETCH after UNANNOUNCE would be coming _from_ the relay on the cell connection.  Your totally right there's a race condition here (GOAWAY may help?) -- all the more reason to say you might receive a FETCH for something you have UNANNOUNCED.

Publishing through the relay starts with publisher sending ANNOUNCE
control message with a `Track Namespace` ({{model-track}}).
The ANNOUNCE enables the relay to know which publisher to forward a
SUBSCRIBE to.  Alternately, the publisher can send one or more PUBLISH messages
to establish subscriptions to tracks.  Relays MAY respond with PUBLISH_OK in

> Forward State=0 until there are known subscribers for new tracks.

Reviewer 3: nit but I think I would flip this around and describe publish first then make annouce be the additional way to do it.

Relays MUST verify that publishers are authorized to publish the content
associated with the set of tracks whose Track Namespace matches the namespace in
an ANNOUNCE, or the Full Track Name in PUBLISH. Where the authorization and
identification of the publisher occurs depends on the way the relay is managed
and is application specific.

A Relay can receive announcements for the same Track Namespace or PUBLISH
messages for the same track from multiple publishers, and it SHOULD respond with

> the same response to each of the publishers.

Reviewer 3: so clearly is can't send the literaly the same response to a publish that was sent to an announce. I think this might just need a bit more detail when talking about response to an announce and to a publish inside the same namespace.  I've always felt the text here was weak. Clearly if one of the request had a bad credential and one had a good credential, we would not send the same response.

Author: @fluffy do you have a suggestion for how you would word this?

Reviewer 3: proposal end the stentence before the ", and it"

When a publisher wants to stop new subscriptions for an announced namespace it
sends an UNANNOUNCE. A subscriber indicates it will no longer subcribe to tracks
in a namespace it previously responded ANNOUNCE_OK to by sending an
ANNOUNCE_CANCEL.

A relay manages sessions from multiple publishers and subscribers, connecting
them based on the track namespace or track name.  Prefix matching is used to
determine which publishers receive a SUBSCRIBE or which subscribers receive a
PUBLISH. For example, a SUBSCRIBE namespace=(foo,bar), track=x message will be
forwarded to the sessions that sent ANNOUNCE namespace=(foo) and ANNOUNCE
namespace=(foo, bar) respectively, but not one that sent ANNOUNCE

> namespace=(foobar).  Relays MUST forward to all publishers or subscribers that

Reviewer 3: I think you mean "annoucer" not subscriber on above line

Author: > Relays MUST forward to all publishers or subscribers

Here?  SUBSCRIBE is forwarded to a publisher (eg: who sent ANNOUNCE) and PUBLISH is sent to a subscriber (wg: who sent SUBSCRIBE_ANNOUNCES).

Reviewer 3: subscribes go to publisher and ...

have a namespace prefix match.  TODO: do we need to say more about this.

> When a relay receives an incoming SUBSCRIBE request that triggers an upstream

Reviewer: ```suggestion
When a relay receives an incoming SUBSCRIBE that triggers an upstream
```

> subscription, it SHOULD send a SUBSCRIBE request to each publisher that has
> announced the subscription's namespace or prefix thereof, unless it already has

Reviewer: ```suggestion
subscription, it SHOULD send a SUBSCRIBE request to each publisher that has
announced the subscription's namespace or prefix thereof, unless it already has
```

Each publisher?

an active subscription for the Objects requested by the incoming SUBSCRIBE
request from all available publishers.  If it already has a matching upstream
subscription in Forward State=0, it SHOULD send a SUBSCRIBE_UDPATE with

> Forward=1 to all publishers.

Reviewer: What if the new subscription is also Forward State=0?

Author: We say elsewhere that when a relay gets SUBSCRIBE forward=0 to should send SUBSCRIBE forward=1 upstream, I was trying to keep this symmetric.

When a relay receives an incoming PUBLISH message, it SHOULD send a PUBLISH

> request to each subscriber that has subscribed to the track's namespace or

Reviewer 2: ```suggestion
request to each subscriber that has subscribed to the track's namespace (via SUBSCRIBE_ANNOUNCES) or
```

> prefix thereof.

Reviewer 2: ```suggestion
prefix or FullTrackName (via SUBSCRIBE) thereof.
```

Author: Should SUBSCRIBE(full track name) should solicit PUBLISH(full track name)?  The SUBSCRIBE will resolve by forwarding upstream.  If you are waiting for a particular track to show up (eg pending SUBSCRIBE), shouldn't you SUBSCRIBE_ANNOUNCES (namespace) ?

> request to each subscriber that has subscribed to the track's namespace or
> prefix thereof.

Author: ```suggestion
request to each subscriber that has subscribed (via SUBSCRIBE_ANNOUNCES)
to the track's namespace or prefix thereof.
```

associated with the set of tracks whose Track Namespace matches the namespace in
an ANNOUNCE, or the Full Track Name in PUBLISH. Where the authorization and
identification of the publisher occurs depends on the way the relay is managed
and is application specific.

A Relay can receive announcements for the same Track Namespace or PUBLISH
messages for the same track from multiple publishers, and it SHOULD respond with
the same response to each of the publishers.

When a publisher wants to stop new subscriptions for an announced namespace it
sends an UNANNOUNCE. A subscriber indicates it will no longer subcribe to tracks
in a namespace it previously responded ANNOUNCE_OK to by sending an
ANNOUNCE_CANCEL.

A relay manages sessions from multiple publishers and subscribers, connecting
them based on the track namespace or track name.  Prefix matching is used to
determine which publishers receive a SUBSCRIBE or which subscribers receive a
PUBLISH. For example, a SUBSCRIBE namespace=(foo,bar), track=x message will be
forwarded to the sessions that sent ANNOUNCE namespace=(foo) and ANNOUNCE
namespace=(foo, bar) respectively, but not one that sent ANNOUNCE

> namespace=(foobar).  Relays MUST forward to all publishers or subscribers that
> have a namespace prefix match.  TODO: do we need to say more about this.

Author: ```suggestion
namespace=(foobar).  Relays MUST forward SUBSCRIBE messages to all publishers
and ANNOUNCE and PUBLISH messages to all subscribers that have a namespace
prefix match.
```

* Too Far Behind - The publisher's queue of objects to be sent to the given
  subscriber exceeds its implementation defined limit.

## PUBLISH {#message-publish}

The publisher sends the PUBLISH control message to initiate a subscription to a
track. The receiver verifies the publisher is authorized to publish this track.

~~~
PUBLISH Message {
  Type (i) = 0x6,
  Length (i),
  Request ID (i),
  Track Namespace (tuple),
  Track Name Length (i),
  Track Name (..),
  Track Alias (i),
  Group Order (8),
  ContentExists (8),
  [Largest (Location),]
  Forward (8),
  Number of Parameters (i),
  Parameters (..) ...,

> }

Reviewer 2: I think we are missing Publisher Priority specification . Today a publisher can suggest its priority in SubscribeOk though.

Author: > Today a publisher can suggest its priority in SubscribeOk though.

Discussed offline, no it can't, but you can file an issue if you want it to?

* Forward: The forward mode for this subscription.  Any value other than 0 or 1
  is a Protocol Violation.  0 indicates the publisher will not transmit any
  objects until the subscriber sets the Forward State to 1. 1 indicates the
  publisher will start transmitting objects immediately, even before PUBLISH_OK.

* Parameters: The parameters are defined in {{version-specific-params}}.

## PUBLISH_OK {#message-publish-ok}

The subscriber sends an PUBLISH_OK control message to acknowledge the successful
authorization and acceptance of a PUBLISH message, and establish a subscription.

~~~
PUBLISH_OK Message
{
  Type (i) = 0x7,
  Length (i),
  Request ID (i),
  Forward (8),
  Subscriber Priority (8),
  Group Order (8),
  Filter Type (i),
  [Start (Location)],
  [EndGroup (i)],
  Number of Parameters (i),
  Parameters (..) ...,
}
~~~
{: #moq-transport-publish-ok format title="MOQT PUBLISH_OK Message"}

* Request ID: The Request ID of the PUBLISH this message is replying to
  {{message-publish}}.

* Forward: The Forward State for this subscription, either 0 (don't
  forward) or 1 (forward).

* Subscriber Priority: The Subscriber Priority for this subscription.

> * Group Order: The Group Order preference for this subscription.

Reviewer 2: ```suggestion
* Group Order: The Group Order preference for this subscription.  This
overwrites the GroupOrder specified PUBLISH.
```

* Request ID: The Request ID of the PUBLISH this message is replying to
  {{message-publish}}.

* Error Code: Identifies an integer error code for failure.

* Reason Phrase: Provides the reason for error.

The application SHOULD use a relevant error code in PUBLISH_ERROR, as defined
below:

|------|---------------------------|
| Code | Reason                    |
|-----:|:--------------------------|
| 0x0  | Internal Error            |
|------|---------------------------|
| 0x1  | Unauthorized              |
|------|---------------------------|
| 0x2  | Timeout                   |
|------|---------------------------|
| 0x3  | Not Supported             |
|------|---------------------------|
| 0x4  | Uninterested              |
|------|---------------------------|

* Internal Error - An implementation specific or generic error occurred.

* Unauthorized - The publisher is not authorized to publish the given
  namespace or track.

* Timeout - The subscription could not be established before an
  implementation specific timeout.

* Not Supported - The endpoint does not support the PUBLISH method.

* Uninterested - The namespace or track is not of interest to the

>   endpoint.

Reviewer 2: Not too sure how can this be used ?

Author: If I SUBSCRIBE_NAMESPACE to foo, and publisher send PUBLISH foo, bar and PUBLISH foo, baz - maybe I only want foo, bar.  I will send PUBLISH_ERROR UNINTERESTED.

  Parameters (..) ...,
}
~~~
{: #moq-transport-publish-ok format title="MOQT PUBLISH_OK Message"}

* Request ID: The Request ID of the PUBLISH this message is replying to
  {{message-publish}}.

* Forward: The Forward State for this subscription, either 0 (don't
  forward) or 1 (forward).

* Subscriber Priority: The Subscriber Priority for this subscription.

* Group Order: The Group Order preference for this subscription.  This
overwrites the GroupOrder specified PUBLISH.

* Filter Type, Start, End Group: See {{message-subscribe-req}}.

* Parameters: Parameters associated with this message.

## PUBLISH_ERROR {#message-publish-error}

The subscriber sends an PUBLISH_ERROR control message for to reject
a subscription initiated by PUBLISH.

~~~
PUBLISH_ERROR Message
{
  Type (i) = 0x8,
  Length (i),
  Request ID (i),
  Error Code (i),
  Reason Phrase Length (i),
  Reason Phrase (..),
}
~~~
{: #moq-transport-publish-error format title="MOQT PUBLISH_ERROR Message"}

* Request ID: The Request ID of the PUBLISH this message is replying to
  {{message-publish}}.

* Error Code: Identifies an integer error code for failure.

* Reason Phrase: Provides the reason for error.

> The application SHOULD use a relevant error code in PUBLISH_ERROR, as defined

Reviewer 3: stylistic issue of no real relevance to the PR. Why is error code a SHOULD ? I have this feeling that you think SHOULD means something different than I do.

Author: I think we sometimes say SHOULD for something we think is a really good idea (send a useful error code), but no one will build enforcement for.  That said, I copied this from SUBSCRIBE_ERROR and I don't think it's adding much value - I could delete it.

* Parameters: The parameters are defined in {{version-specific-params}}.

The publisher will respond with SUBSCRIBE_ANNOUNCES_OK or
SUBSCRIBE_ANNOUNCES_ERROR.  If the SUBSCRIBE_ANNOUNCES is successful, the

> publisher will forward any matching ANNOUNCE and PUBLISH messages to the

Author: Make it clear that existing ANNOUNCE / subscriptions are forwarded to new SUBSCRIBE_ANNOUNCES
